Sources/ReferenceResolver.swift

```diff
@@ -0,0 +1,87 @@
+import ObjectMapper
+
+protocol ResolvableType: Builder {
+    static var path: String { get }
+}
+
+private let kSchemaBuilderReferenceResolver = ReferenceResolver<SchemaBuilder>()
+extension SchemaBuilder: ResolvableType {
+    static var path: String { return "definitions" }
+    static var resolver: ReferenceResolver<SchemaBuilder> { return kSchemaBuilderReferenceResolver }
+}
+
+private let kParameterBuilderReferenceResolver = ReferenceResolver<ParameterBuilder>()
+extension ParameterBuilder: ResolvableType {
+    static var path: String { return "parameters" }
+    static var resolver: ReferenceResolver<ParameterBuilder> { return kParameterBuilderReferenceResolver }
+}
+
+private let kResponseBuilderReferenceResolver = ReferenceResolver<ResponseBuilder>()
+extension ResponseBuilder: ResolvableType {
+    static var path: String { return "responses" }
+    static var resolver: ReferenceResolver<ResponseBuilder> { return kResponseBuilderReferenceResolver }
+}
+
+
+class ReferenceResolver<T: ResolvableType> {
+    enum ResolverError: Error {
+        case invalidContext
+        case invalidReference
+        case invalidPath
+        case unsupportedReference
+        case unresolvedReference
+    }
+
+    private class Context<C> {
+        var cachedReferences = [String: C]()
+        var resolvingReferences = [String: Bool]()
+    }
+
+    private var context: Context<T.Building>?
+
+    func setup() {
+        self.context = Context<T.Building>()
+    }
+
+    func resolve(_ swagger: SwaggerBuilder, pointer: Pointer<T>) throws -> Structure<T.Building> {
+        guard let context = self.context else {
+            throw ResolverError.invalidContext
+        }
+
+        let components = pointer.path.components(separatedBy: "/")
+        guard components.count == 3 && components[0] == "#" && components[1] == T.path else {
+            throw ResolverError.invalidPath
+        }
+
+        let name = components[2]
+        let referencedBuilder: T?
+        switch T.path {
+        case "definitions": referencedBuilder = swagger.definitions[name] as? T
+        case "parameters": referencedBuilder = swagger.parameters[name] as? T
+        case "responses": referencedBuilder = swagger.responses[name] as? T
+        default: throw ResolverError.unsupportedReference
+        }
+
+        guard let builder = referencedBuilder else {
+            throw ResolverError.unresolvedReference
+        }
+
+        if let cached = context.cachedReferences[name] {
+            return Structure(name: name, type: cached)
+        }
+
+        if let resolving = context.resolvingReferences[name], resolving {
+            return Structure(name: name, type: nil)
+        }
+
+        context.resolvingReferences[name] = true
```

Comment:
Could you add a short comment describing how this works / why you set this temporarily? It took me a minute to process why this is set to true here and then nil after building is complete.